src/cmap/auth/gssapi.ts

```diff
@@ -21,65 +31,159 @@ export class GSSAPI extends AuthProvider {
       return callback(Kerberos['kModuleError']);
     }
 
-    const username = credentials.username;
-    const password = credentials.password;
-    const mechanismProperties = credentials.mechanismProperties;
-    const gssapiServiceName =
+    const { username, password, mechanismProperties } = credentials;
+    const serviceName =
       mechanismProperties['gssapiservicename'] ||
       mechanismProperties['gssapiServiceName'] ||
       'mongodb';
+    const canonicalizeHostName =
+      typeof mechanismProperties.gssapiCanonicalizeHostName === 'boolean'
+        ? mechanismProperties.gssapiCanonicalizeHostName
+        : false;
 
-    const MongoAuthProcess = Kerberos.processes.MongoAuthProcess;
+    performGssapiCanonicalizeHostName(
+      canonicalizeHostName,
+      host,
+      (err?: Error | MongoError, host?: string) => {
+        if (err) return callback(err);
 
-    const authProcess = new MongoAuthProcess(host, port, gssapiServiceName, mechanismProperties);
+        const initOptions = {};
+        if (password != null) {
+          Object.assign(initOptions, { user: username, password: password });
+        }
 
-    authProcess.init(username, password, err => {
-      if (err) return callback(err, false);
-      authProcess.transition('', (err, payload) => {
-        if (err) return callback(err, false);
-
-        const command = {
-          saslStart: 1,
-          mechanism: 'GSSAPI',
-          payload,
-          autoAuthorize: 1
-        };
+        Kerberos.initializeClient(
+          `${serviceName}${process.platform === 'win32' ? '/' : '@'}${host}`,
+          initOptions,
+          (err: string, client: KerberosClient): void => {
+            if (err) return callback(new Error(err));
+            if (client == null) return callback(new Error('null gssapi client'));
+            this[kGssapiClient] = client;
+            callback(undefined, handshakeDoc);
+          }
+        );
+      }
+    );
+  }
 
-        connection.command('$external.$cmd', command, (err, doc) => {
-          if (err) return callback(err, false);
+  auth(authContext: AuthContext, callback: Callback): void {
+    const { connection, credentials } = authContext;
+    if (credentials == null) return callback(new Error('credentials required'));
```

Comment:
Let's not ever use plain `Error`s, can we change these to `MongoError` or `TypeError` depending on context? Also, in this particular case I think if at this point no credentials are found then we should just `callback()` without error